### ISSUE TITLE
Huge overhead costs when minimize is called many times in same python session

### ISSUE THREAD

**Issue by Selmaan** [GPflow/GPflow#798]
This issue arises for me when I am attempting to fit many different models in one session, but the issue can be reproduced with a minimal example. Assume I have defined a kernel above, as well as loaded data X and Y and defined inducing points Z:

```python
model = gpflow.models.SGPR(X, Y, kernel, Z)
opt = gpflow.train.ScipyOptimizer()
for n in np.arange(100):
    opt.minimize(model, maxiter = 3, anchor=True)
    #opt.minimize(model, maxiter = 3, anchor=False)
```

What is odd is that when `anchor=False` there is no slowdown observed, but when `anchor=True` each loop iteration is slower than the last. I'd be glad to post more specifics on how I have constructed the model if that's relevant here.

**Comment by hughsalimbeni** [comment]
This is a noted problem, and is due to expensive assigns with interactions between python/tensorflow. If you don't use `anchor=True` then that's fine as long as you keep track of the session and don't rely on the python variables. You can see an example of avoiding this overhead here https://github.com/GPflow/GPflow/blob/hugh/advanced_usage_notebook/doc/source/notebooks/advanced_usage.ipynb

**Comment by Selmaan** [comment]
Thanks! that sounds like exactly what I was looking for

**Comment by Selmaan** [comment]
Actually, I'm re-opening this issue, as setting `anchor=False` reduces the overhead substantially but does not totally eliminate this issue. This is using the 'feed_dict' solution proposed above (which also was an improvement, thanks).

As an aside, it appears that tensors which are fixed can be passed in the 'feed_dict' without issue. However, when passing in trainable parameters to the 'feed_dict' for an optimizer.minimize call (i.e. `opt.minimize(model, maxiter = 3, anchor=False, feed_dict=feed_dict)` ), this appears to break optimizer functionality.

**Comment by st--** [comment]
@Selmaan opt.minimize() adds operations to the tensorflow graph. By default, a single graph is shared, but when you loop over model fits, this can significantly slow down computations (and increases the memory usage). You can use `opt_tensor = opt.make_optimize_tensor(model)` and then explicitly call `session.run(opt_tensor, feed_dict=feed_dict)`.

**Comment by nrontsis** [comment]
@st-- wouldn't it be possible to split `opt.minimize()` into something like `opt.build()` and `opt.run()`?

**Comment by st--** [comment]
@nrontsis well, `make_optimize_tensor` and `make_optimize_action` do the building already; minimize takes care of running that as well as callbacks. I'm not quite sure what you are proposing - add a kwarg to minimize() that lets you reuse an action?

NB- an important point to keep in mind, when you change the "trainable" state of any model parameter, you will need to construct a new optimizer (sub)graph ... as there'll need to be different ops when only some (or additional) params get updated.

**Comment by kyr-pol** [comment]
Hi @st-- I tried to use the `opt.make_optimize_tensor(model)` trick, but I am getting an error: 

```
TypeError: Fetch argument <gpflow.training.external_optimizer.ScipyOptimizerInterface object at 
0x11e8862b0> has invalid type <class 'gpflow.training.external_optimizer.ScipyOptimizerInterface'>, 
must be a string or Tensor. (Can not convert a ScipyOptimizerInterface into a Tensor or Operation.)
```

Am I missing something obvious here, or did something change recently? 
MWE:
```python
kernel = gpflow.kernels.RBF(input_dim=1)
model = gpflow.models.GPR(X1, Y1, kernel)
model.compile()
opt = gpflow.train.ScipyOptimizer()
opt_tensor = opt.make_optimize_tensor(model)
session = tf.Session()
for n in np.arange(5):
    session.run(opt_tensor, feed_dict={X1: X, Y1: Y})
```
where X1 and Y1 are placeholders and X,Y some data. 
I am using gpflow 1.2.0 and tensorflow 1.11.0.

**Comment by nrontsis** [comment]
@kyr-pol I think that [this commit](https://github.com/nrontsis/PILCO/commit/a88e6e1bcd6b2c042258cc5246810781999a852e) addresses your/our issue.

**Comment by nrontsis** [comment]
@st-- I think that, currently, repeated (say 10) efficient calls to model optimization can be made like this:
```
optimizer = gpflow.train.ScipyOptimizer(method='L-BFGS-B')
optimizer.minimize(model)
for i in range(9):
    optimizer._optimizer.minimize(session=optimizer._model.enquire_session(None),
        feed_dict=optimizer._gen_feed_dict(optimizer._model, None),
        step_callback=None)
```
My suggestion is to change the user interface to something like:
```
optimizer = gpflow.train.ScipyOptimizer(method='L-BFGS-B')
optimizer.build(model)
for i in range(10):
    optimizer.minimize()
```

**Comment by st--** [comment]
@kyr-pol I'm sorry, when I answered I forgot that you were using ScipyOptimizer not one of the TensorflowOptimizers such as Adam ... what I said applies to the latter. For ScipyOptimizer, unfortunately it's more complicated (because all data need to be handled manually); @awav made a hack to get similar optimizer-reusability in a separate branch (https://github.com/GPflow/GPflow/tree/awav/reuse-scipyoptimizer), but using it is not trivial (@awav could you post the instructions here?).

@nrontsis if you can get it to work you're very welcome to make a PR to GPflow! :-)

**Comment by st--** [comment]
@awav is there anything blocking us from merging this into develop?

**Comment by st--** [comment]
@awav ping

**Comment by FredericSauv** [comment]
Hi,
What is the status of this thread (I can see some recent  commits related to it). And what is the best way now to repeatedly use a scipy optimizer?
Cheers

**Comment by st--** [comment]
@FredericSauv this was fixed by #1011, which is part of the latest GPflow release, version 1.5.0.
`ScipyOptimizer().minimize(model)` now caches the optimisation tensor, so you can simply call `minimize()` as many times as you want. You can change model.X, model.Y etc in between calls, but I *think* if you change the "trainable" status of parameters you may need to construct a new optimizer - @awav should be able to confirm. There is an open issue to improve the documentation around this, #1036.

**Comment by FredericSauv** [comment]
Thanks!
